### ISSUE TITLE
Store#onClose seems to leak memory

### ISSUE THREAD

**Issue by vlsi** [jlink/jqwik#290]
## Testing Problem

See https://jqwik.net/docs/current/user-guide.html#temporaryfilehook

The issue is that `tempFileStore.onClose(file -> file.delete());` would be executed multiple times, so it would effectively add the same callback again and again.

```java
    @Override
    public Optional<ParameterSupplier> resolve(ParameterResolutionContext parameterContext, LifecycleContext lifecycleContext) {
        if (parameterContext.typeUsage().isOfType(File.class)) {
            return Optional.of(ignoreTry -> getTemporaryFileForTry());
        }
        return Optional.empty();
    }
  
    private File getTemporaryFileForTry() {
        Store<File> tempFileStore = Store.getOrCreate(STORE_IDENTIFIER, Lifespan.TRY, this::createTempFile);
        tempFileStore.onClose(file -> file.delete());
        return tempFileStore.get();
    }
```

## Suggested Solution

`onClose` callback registration should be moved to store creation.

For instance:

```kotlin
val tempFileStore = createStore(STORE_IDENTIFIER, Lifespan.TRY) { // this: StoreBuilder<T>
    onClose { it.delete() } // StoreBuilder#onClose(Consumer<T>)
    createTempFile() // creates initial value
} // returns Store<T>
```

**Comment by vlsi** [comment]
Draft implementation:

```kotlin
import net.jqwik.api.lifecycle.CannotFindStoreException
import net.jqwik.api.lifecycle.Lifespan
import net.jqwik.api.lifecycle.Store
import java.lang.IllegalArgumentException
import java.util.function.Consumer
import java.util.function.Supplier

class StoreBuilder<T> {
    internal val cleanups = mutableListOf<Consumer<T>>()
    internal lateinit var initialValue: Supplier<T>

    fun onClose(cleanup: Consumer<T>) {
        cleanups += cleanup
    }

    fun initialValue(initialValue: Supplier<T>) {
        this.initialValue = initialValue
    }
}

// See https://github.com/jlink/jqwik/issues/290
fun <T> createStore(
    identifier: Any,
    lifespan: Lifespan,
    configure: StoreBuilder<T>.() -> Unit
): Store<T> =
    try {
        Store.get<T>(identifier).apply {
            if (lifespan != lifespan()) {
                throw IllegalArgumentException("Can't create store $identifier with lifespan $lifespan, as existing one has lifespan ${lifespan()} ")
            }
        }
    } catch (e: CannotFindStoreException) {
        val params = StoreBuilder<T>().apply(configure)
        Store.create(identifier, lifespan, params.initialValue).apply {
            for (cleanup in params.cleanups) {
                onClose(cleanup)
            }
        }
    }
```

Sample usage:

```kotlin
    private fun connect(): Connection =
        createStore<Connection>(DbConnectionHook::class, Lifespan.RUN) {
            onClose { it.close() }
            initialValue { TestUtil.openDB() } // <-- it is sad that initial value should be wrapped into initialValue function
        }.get()
```

Unfortunately, the current `Store.create*` API does not allow to make usages as follows:

```kotlin
        createStore<Connection>(DbConnectionHook::class, Lifespan.RUN) {
            onClose { it.close() }
            TestUtil.openDB() // <-- this is not possible since Store initializer does not provide access to the store itself
        }.get()
```

**Comment by jlink** [comment]
@vlsi Thanks for bringing up this discussion!

In the case you describe the deleting the file will only be attempted once. 
You can see that by adding a logging command in the callback:

```java
tempFileStore.onClose(file -> {
	System.out.println("## delete: " + file.getName());
	file.delete();
});
```

The reason is that onClose-callbacks are stored in a HashSet. Combined with the way how the JVM creates instances of lambda expressions, only the first callback will actually be added to the set of callbacks.

This behaviour changes, however, as soon as the lambda's closure will comprise an outside variable:

```java
int randomInt = new Random().nextInt();
tempFileStore.onClose(file -> {
	System.out.println("## delete: " + file.getName());
	System.out.println("## random: " + randomInt);
	file.delete();
});
```

But then, you may argue, it actually _is_ a different callback and should be added to the set of callbacks.

To be frank, your suggestion to add onClose callbacks during store initialization is logically sounder. I am on the fence, however, if this breaking change is worth it. On the other hand, `onClose` is marked as `EXPERIMENTAL` so I guess not too many jqwik users use it.

How would you judge it?

**Comment by jlink** [comment]
One more thing...

I think your intended API can be implemented like that:

```kotlin
class StoreBuilder<T> {
    internal val cleanups = mutableListOf<Consumer<T>>()

    fun onClose(cleanup: Consumer<T>) {
        cleanups += cleanup
    }
}

fun <T> createStore(
    identifier: Any,
    lifespan: Lifespan,
    configure: StoreBuilder<T>.() -> T
): Store<T> =
    try {
        Store.get<T>(identifier).apply {
            if (lifespan != lifespan()) {
                throw IllegalArgumentException("Can't create store $identifier with lifespan $lifespan, as existing one has lifespan ${lifespan()} ")
            }
        }
    } catch (e: CannotFindStoreException) {
        val builder = StoreBuilder<T>()
        val initialValue = builder.configure();
        Store.create(identifier, lifespan, { initialValue }).apply {
            for (cleanup in builder.cleanups) {
                onClose(cleanup)
            }
        }
    }
```

Example usage: 

```kotlin
val store: Store<String> =
    createStore("hello", Lifespan.TRY) {
        onClose { println("Closing $it") }
        "hello"
    }

@Property(tries = 10)
fun test() {
    println(store.get())
}
```

I wonder if it looks as nice in plain Java, though.

**Comment by vlsi** [comment]
>Combined with the way how the JVM creates instances of lambda expressions, only the first callback will actually be added to the set of callbacks.

I believe the spec says lambdas have no identity, so it is wrong to assume they can be added to a hashset :-/

Gradle relied on a similar properties for lambdas, and AFAIK, they drift away from that, and they try to avoid relying on the lambda identity.

It does not feel right that the callbacks are added on each store.get

**Comment by jlink** [comment]
I share your feeling. You convinced me. Will change the API.

**Comment by vlsi** [comment]
It looks like builder pattern for configuring the Store-to-be-created feels right here.

For instance:

```java
StoreConfig sc = StoreConfig.builder("hello")
    .lifespan(Lifespan.TRY)
    .initialValue(() -> new ...)
    .onClose((value) -> value.close())
    .build();

Store s = Store.getOrCreate(sc);
``

It would unify the current Store.get/create/... methods, it would avoid duplicate subscriptions for onClose, etc.

I see the API for the first time, so I have no idea if the suggestions make sense though :)

**Comment by vlsi** [comment]
>I wonder if it looks as nice in plain Java, though.

Frankly speaking, it is a tough challenge for me.
I find Kotlin is relatively easy to introduce, especially for testing purposes.

Recently @dalexandrov added Kotlin convenience methods to Helidon, and the key principle was "whenever you see Builder in Java, it could be refined with a lambda-with-receiver in Kotlin".
See https://medium.com/helidon/kotlin-helidon-9af5c35c830a

So it looks like "builders in Java" are "the best Java can have" and the builders could be refined for Kotlin 😉

**Comment by jlink** [comment]
Having a good API for both Java and Kotlin is tricky. I tend to make the Java API good first and if necessary add convenience method(s) for Kotlin. Before I saw your suggestion with a builder, I was thinking of something like that:

```java
interface StoreInitializer<T> {
	void onClose(Consumer<T> onCloseHandler);
}

class Store {
	static <T> Store<T> create(Object identifier, Lifespan lifespan, Function<StoreInitializer<T>, T> initializer) {...}
	
	static <T> Store<T> getOrCreate(Object identifier, Lifespan lifespan, Function<StoreInitializer<T>, T> initializer) {...}
}
```

The Kotlin convenience function would then look like

```kotlin
// Provide (create or get) a Store object for type T
fun <T> provideStore(
    identifier: Any,
    lifespan: Lifespan,
    initialize: StoreInitializer<T>.() -> T
): Store<T> = ...
```

The advantage I see: It's just overloading the existing API and it requires fewer internal cogs. 
The builder implementation would have to be rather complicated to ensure that `identifier`, `lifespan` and `initialValue` are only specified once.

**Comment by jlink** [comment]
Or, keeping the builder but skipping the config:

```java
Store<T> sc = Store.builder("hello")
    .lifespan(Lifespan.TRY)
    .initialValue(() -> new ...)
    .onClose((value) -> value.close())
    .build();  // Alternative: getOrBuild() or provide()
```

Would a builder suggest a default `Lifespan`?

**Comment by jlink** [comment]
The Java part is done. I went with:

```java
class Store {

	@API(status = EXPERIMENTAL, since = "1.6.3")
	interface Initializer<T> {
		Initializer<T> onClose(Consumer<T> onCloseCallback);
		void initialValue(T value);
	}

	// Overloads the existing getOrCreate method
	static <T> Store<T> getOrCreate(Object identifier, Lifespan lifespan, Consumer<Initializer<T>>  initialize) { 
		//...
	}
}
```

As for the Kotlin "convenience" function I see a few variations:
- Have all of `getLifecycleStore`, `createLifecycleStore`, `getOrCreateLifecycleStore` or a single 
  `provideLifecycleStore` (better name suggestion welcome).
- Should the initialization context have an InitialValue method:
   ```kotlin
  interface StoreInitializationContext<T> {
	void onClose(Consumer<T> onCloseHandler) // can be called several times
	void initialValue(T initialValue) // can be called at most once
  }
  ```
  which would allow reusing `Store.Initializer` or should the initialization block just return the initial value:
  ```kotlin
  fun <T> provideLifecycleStore(
      identifier: Any,
      lifespan: Lifespan,
      initialize: StoreInitializationContext<T>.() -> T
  ): Store<T> = ...
  ```

@vlsi What would you prefer?

**Comment by vlsi** [comment]
By the way, if you use `fun interface` in Kotlin, then Java would be able to use it like `Consumer`, and Kotlin would be able to use receiver.

Sample:

```kotlin
fun interface MyAction {
    fun Int.onSucceed(arg: String): Unit
}
```

Java

```java
public class Test {
    public static void test(MyAction action) {
    }

    public static void main(String[] args) {
        test((intReceiver, stringArg) -> {
        });
    }
}
```

```kotlin
fun test() {
    Test.test { arg -> // arg: String
        // this: Int
    }
}
```

**Comment by vlsi** [comment]
@jlink , is there a real need for `Store.getOrCreate`?

For instance, what if different code branches happen to create a store with the same ID?
jqwik would happily overwrite the data.

What if the API was as follows?

```
Store.get(id); // throw if store does not exist
Store.getOrNull(id); // returns null if the store is missing
Store.create(id, lifetime, ...); // creates the store or throws if the same-named store exists no matter what the lifetime or whatever
```

**Comment by vlsi** [comment]
I know Gradle tasks have only "get" and "create" APIs, and it works for 99.99% of the cases.
There are rare requests for "get or create", however, it looks like the users should not need `Store.getOrCreate`.

**Comment by jlink** [comment]
> @jlink , is there a real need for `Store.getOrCreate`?
> 
> For instance, what if different code branches happen to create a store with the same ID? jqwik would happily overwrite the data.

`getOrCreate()` is a convenience and it's the one I mostly use. The reason for that is that many hook methods can be called several times during the same try. And only the first time the store should actually be generated.

In order to prevent overwriting a store with the same ID by coincidence, the ID is usually a composed one using the hook/extension class plus a name describing what the store contains, e.g. `Tuple.of(MyHookClass.class, "myObjects")`.

It's almost the other way round, that create() and get() are seldom needed. I think there's is only one usage within jqwik code itself (apart from tests and delegation).

**Comment by vlsi** [comment]
>The reason for that is that many hook **methods can be called several times during the same try**

What do you think of the following?
1. "before try start hook" -> `create store`
2. "parameter resolve hook" -> `get`

**Comment by jlink** [comment]
> What do you think of the following?
> 
>     1. "before try start hook" -> `create store`
> 
>     2. "parameter resolve hook" -> `get`

That can sometimes work, sometimes not. There are, for example, global hooks that need a store if (and only if) their functionality is needed, e.g. statistics. Creating the store only if needed saves resources.

**Comment by jlink** [comment]
BTW, I just changed the Java API so that `Store.create` and `Store.getOrCreate` take an option 4th parameter `Consumer<T> onClose`. After having looked at the `Initializer` and `Builder` approaches for a few days my impression was that they don't fit in.

That said, the jury is still out on the best Kotlin API for creating/providing stores.

**Comment by vlsi** [comment]
> Creating the store only if needed saves resources

Ok, then:

1)  "before try start hook"  -> register store
2) "parameter resolve hook" -> get

The crucial part is that `get(...)` should not contain "lifecycle" parameter.

**Comment by jlink** [comment]
> > Creating the store only if needed saves resources
> 
> Ok, then:
> 
>     1. "before try start hook"  -> register store (e.g. https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:how_do_i_reference_a_task)
> 
>     2. "parameter resolve hook" -> get
> 
> 
> The crucial part is that `get(...)` should not contain "lifecycle" parameter.

What's the difference between "create store" and "register store" in your opinion? The value in the store will only be initialized on first access anyway.

The most prominent example where the first access to the store is somewhat unknown is jqwik's Spring extension: https://github.com/jlink/jqwik-spring/blob/main/src/main/java/net/jqwik/spring/JqwikSpringExtension.java.

**Comment by jlink** [comment]
It just occurred to me that the onClose parameter could be completely eliminated by going the Jupiter round and cleaning up stored objects if they implement `AutoCloseable`.

**Comment by vlsi** [comment]
> onClose parameter could be completely eliminated by going the Jupiter round and cleaning up stored objects if they implement

It might be a bit surprising if the user puts something like `Connection`, and it gets closed unexpectedly because the engine "does automatic close for `AutoCloseable`".

The workaround would probably be wrapping the value into `Optional<T>` (or another box-like class): `Store<Optional<Connection>>` and then use `.get().get()`.

A similar discussion was here: https://github.com/junit-team/junit5/issues/1555

Having a dedicated interface `CloseableResource` might work slightly better: it is the user's choice if the thing should be closed or not.

----

However, I have no strong position here, as my set of use-cases is too limited for now.

**Comment by jlink** [comment]
Introduced interface `Store.CloseOnReset`:

```java
interface Store {
	interface CloseOnReset {
		void close() throws Exception;
	}
}
```

Removed all methods in `Store` with `onClose` parameter.

Convenience method for Kotlin seems no longer necessary.

Available in 1.6.3-SNAPSHOT.

**Comment by jlink** [comment]
I think it's ready for upcoming release 1.6.3. 

Please reopen if not ready for the coming release.
Please create a new issue for enhancements.
